README.md
# react-leaflet-dialog

[![version](https://img.shields.io/npm/v/react-leaflet-dialog.svg?style=plastic)](http://npm.im/react-leaflet-dialog)
[![react-leaflet compatibility](https://img.shields.io/npm/dependency-version/react-leaflet-dialog/peer/react-leaflet.svg?style=plastic)](https://github.com/mhasbie/react-leaflet-dialog)
[![travis build](https://img.shields.io/travis/mhasbie/react-leaflet-dialog.svg?style=plastic)](https://travis-ci.org/mhasbie/react-leaflet-dialog)
[![dependencies](https://img.shields.io/david/mhasbie/react-leaflet-dialog.svg?style=plastic)](https://david-dm.org/mhasbie/react-leaflet-dialog)
[![peer dependencies](https://img.shields.io/david/peer/mhasbie/react-leaflet-dialog.svg?style=plastic)](https://david-dm.org/mhasbie/react-leaflet-dialog?type=peer)
[![issues](https://img.shields.io/github/issues/mhasbie/react-leaflet-dialog.svg?style=plastic)](https://github.com/mhasbie/react-leaflet-dialog/issues)
[![downloads](https://img.shields.io/npm/dt/react-leaflet-dialog.svg?style=plastic)](http://npm-stat.com/charts.html?package=react-leaflet-dialog&from=2018-01-01)
[![MIT License](https://img.shields.io/npm/l/react-leaflet-dialog.svg?style=plastic)](http://opensource.org/licenses/MIT)


React wrapper of [Leaflet.Dialog](https://github.com/NBTSolutions/Leaflet.Dialog) for [react-leaflet](https://github.com/PaulLeCam/react-leaflet).

A dialog modal window that is resizable and positionable on the map.

*Tested with Leaflet 1.4.0 and React-Leaflet 1.9.1, React-Leaflet 2.2.0*


## Demos

| Version	| Demo	|
| ---		| ---	|
| react-leaflet@1.9.1 | [`CodePen`](https://codepen.io/m_hasbie/full/omePOP/) |
| react-leaflet@2.x | [`CodePen`](https://codepen.io/m_hasbie/full/gqxBKr/) |



## Installation

### Install via NPM

```bash
npm install react-leaflet-dialog --save
```

Include font-awesome css to your project.

```html
<link href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
```

## Usage example for `react-leaflet` **v1**


```javascript
import { Map, TileLayer } from 'react-leaflet';
import Dialog from 'react-leaflet-dialog';
		
<Map center={[101.483459, 2.938926]} zoom={12}>
  <TileLayer
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
    attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
  />

  <Dialog ref={(ref) => { this.dialog = ref; }} id="dialog1">
	<div>Dialog content.</div>
  </Dialog>
</Map>
```


## Usage example for `react-leaflet` **v2**


```javascript
import { Map, TileLayer, withLeaflet } from 'react-leaflet';
import DialogDefault from 'react-leaflet-dialog';

// wrap `Dialog` component with `withLeaflet` HOC
const Dialog = withLeaflet(DialogDefault);
		
<Map center={[101.483459, 2.938926]} zoom={12}>
  <TileLayer
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
    attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
  />

  <Dialog ref={(ref) => { this.dialog = ref; }} id="dialog1">
	<div>Dialog content.</div>
  </Dialog>
</Map>
```

## Options

Any props passed to Dialog are passed down to Leaflet.Dialog.

Refer [Leaflet.Dialog options](https://github.com/NBTSolutions/Leaflet.Dialog#options) for a complete list of options supported.

| Property | Type | Default | Description
| --- | --- | --- | ---
| size | [Array][width, height] | [ 300, 300 ] | An array of the starting width and height values (in pixels).
| minSize | [Array][width, height] | [ 100, 100 ] | An array with the minimum resize width and height (in pixels).
| maxSize | [Array][width, height] | [ 350, 350 ] | An array with the maximum resize width and height (in pixels).
| anchor | [Array][top, left] | [ 250, 250 ] | The starting point anchor location (from the upper left) in pixels.
| position | [String] | 'topleft' | The leaflet control div to place the modal inside, as a starting reference point.
| initOpen | [Boolean] | true | Whether or not to initialize in an open state.

## Methods

| Method | Returns | Example | Description
| --- | --- | --- | ---
| open() | this | dialog.open(); | Opens the dialog window.
| close() | this | dialog.close(); | Closes the dialog window.
| destroy() | this | dialog.destroy(); | Removes the dialog box from the window.
| setLocation( [Array][top, left] anchor location ) | this | dialog.setLocation( [ 15, 15 ] ); | Move the dialog box to the specified pixel location ( Relative to the 'position' option )
| setSize( [Array][width, height] size ) | this | dialog.setSize( [ 150, 150 ] ); | Resize the dialog window to the specified width and height.
| setContent( [String | node] content ) | this | dialog.setContent("<p>Hello! Welcome to your nice new dialog box!</p>"); | Set new content for the dialog.
| lock() | this | dialog.lock(); | Hides all visible dialog window controls.
| unlock() | this | dialog.unlock(); | Re-instates all dialog window controls.
| freeze() | this | dialog.freeze(); | Hides all visible dialog window movement/sizing controls.
| unfreeze() | this | dialog.unfreeze(); | Re-instates all dialog window movement/sizing controls.
| hideClose() | this | dialog.hideClose(); | Hides the closing 'x'
| showClose() | this | dialog.showClose(); | Shows the closing 'x'
| hideResize() | this | dialog.hideResize(); | Hides the resizing symbol
| showResize() | this | dialog.showResize(); | Shows the resizing symbol

### Example

```javascript

...

openDialog() {
  this.dialog.open();
}

closeDialog() {
  this.dialog.close();
}

destroyDialog() {
  this.dialog.destroy();
}

updateDialogContent() {
  this.dialog.setContent("<div>This is the new dialog content.</div>");
}

render() {
  return (
	<Map center={[101.483459, 2.938926]} zoom={12}>
	  <TileLayer
		url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
		attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
	  />

	  <Dialog ref={(ref) => { this.dialog = ref; }} id="dialog1" />
	</Map>
  );
}

...

```

## Events

| Event | Data | Description
| --- | --- | ---
| onOpened | this | Fired when open() is called.
| onClosed | this | Fired when close() is called or when the 'x' is clicked.
| onDestroyed | this | Fired when destroy() is called.
| onLocked | this | Fired when lock() is called.
| onUnlocked | this | Fired when unlock() is called.
| onFrozen | this | Fired when freeze() is called.
| onUnfrozen | this | Fired when unfreeze() is called.
| onUpdated | this | Fired when contents are set or when added to the map.
| onResizeStart | this | Fired when the resize button is clicked.
| onResizing | this | Fired continuously as the resize button is dragged.
| onResizeEnd | this | Fired when the resize button click is released.
| onMoveStart | this | Fired when the move button is clicked.
| onMoving | this | Fired continuously as the move button is dragged.
| onMoveEnd | this | Fired when the move button click is released.
| onCloseHidden | this | Fired when hideClose() is called.
| onCloseShown | this | Fired when the showClose() is called.
| onResizeHidden | this | Fired when hideResize() is called.
| onResizeShown | this | Fired when the showResize() is called.

# Credits
Credits goes to [NBTSolutions](https://github.com/NBTSolutions) and all the [contributors](https://github.com/NBTSolutions/Leaflet.Dialog/graphs/contributors) for the original work.

# License

MIT License


CHANGELOG.md
react-leaflet-dialog Changelog
=========================


## 2.0.0 (2019-02-03)

* Update dependencies
* Update README
* Add example
* Add unit test


__tests__/Dialog.test.js
/* global describe, expect, it, jest */
import React, { createRef, Component } from 'react';
import { mount } from './enzyme';
import 'jest-enzyme';

import { Map, TileLayer, withLeaflet } from 'react-leaflet';
import DialogDefault from '../dist/react-leaflet-dialog.min.js';
const Dialog = withLeaflet(DialogDefault);

import DialogContainer from './DialogContainer';

describe('Dialog', () => {

	it('Render dialog box with string content', () => {

		const dialogRef = createRef();
		const mapOptions = {
			center: [2.935403, 101.448205],
			zoom: 3,
			minZoom: 1,
			maxZoom: 22,
		};
		const wrapper = mount(
			<div>
				<Map {...mapOptions}>
					<TileLayer url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" />
					<Dialog ref={dialogRef} anchor={[50, 50]} id="testDialog" >
						{ '<div>This is the dialog content.</div>' }
					</Dialog>
				</Map>
			</div>
		);
		
		expect(wrapper).not.toBeEmptyRender();
		expect(dialogRef.current.leafletElement.options.anchor).toEqual([50, 50]);
	});
	
	it('Open the dialog box when open() is called', () => {
		
		const wrapper = mount(<DialogContainer initOpen={false} />);
		const spyDialogFunction = jest.spyOn(wrapper.instance().dialog, 'open');
		
		wrapper.instance().open();
		
		expect(spyDialogFunction).toHaveBeenCalled();
		expect(wrapper).not.toBeEmptyRender();
		expect(wrapper.render().find('div.leaflet-control-dialog').get(0).attribs.style).not.toContain('visibility: hidden');
	});
	
	it('Close the dialog box when close() is called but not destroy it', () => {
		
		const wrapper = mount(<DialogContainer />);
		const spyDialogFunction = jest.spyOn(wrapper.instance().dialog, 'close');
		
		wrapper.instance().close();
		
		expect(spyDialogFunction).toHaveBeenCalled();
		expect(wrapper).not.toBeEmptyRender();
		expect(wrapper.render().find('div.leaflet-control-dialog').get(0).attribs.style).toContain('visibility: hidden');
	});
	
	it('Destroy the dialog box when destroy() is called', () => {
		
		const wrapper = mount(<DialogContainer />);
		const spyDialogFunction = jest.spyOn(wrapper.instance().dialog, 'destroy');
		
		wrapper.instance().destroy();
		
		expect(spyDialogFunction).toHaveBeenCalled();
		expect(wrapper).not.toBeEmptyRender();
		expect(wrapper.render().find('div.leaflet-control-dialog').get(0)).toBeUndefined();
	});
	
	it('Set new content when setContent() is called', () => {
		
		const wrapper = mount(<DialogContainer />);
		const spyDialogFunction = jest.spyOn(wrapper.instance().dialog, 'setContent');
		
		wrapper.instance().setContent('<div>New dialog content.</div>');
		
		expect(spyDialogFunction).toHaveBeenCalledWith('<div>New dialog content.</div>');
		expect(wrapper).not.toBeEmptyRender();
		expect(wrapper.render().find('#dialog-contents-testDialog').html()).toEqual('<div>New dialog content.</div>');
	});
})

__tests__/DialogContainer.js
/* global describe, expect, it, jest */
import React, { createRef, Component } from 'react';

import { Map, TileLayer, withLeaflet } from 'react-leaflet';
import DialogDefault from '../dist/react-leaflet-dialog.min.js';
const Dialog = withLeaflet(DialogDefault);

export default class DialogContainer extends Component {
	constructor(props) {
		super(props);
		this.open = this.open.bind(this);
		this.close = this.close.bind(this);
		this.destroy = this.destroy.bind(this);
		this.setContent = this.setContent.bind(this);
		this.setLocation = this.setLocation.bind(this);
		this.setSize = this.setSize.bind(this);
		this.lock = this.lock.bind(this);
		this.unlock = this.unlock.bind(this);
		this.freeze = this.freeze.bind(this);
		this.unfreeze = this.unfreeze.bind(this);
		this.hideClose = this.hideClose.bind(this);
		this.showClose = this.showClose.bind(this);
		this.hideResize = this.hideResize.bind(this);
		this.showResize = this.showResize.bind(this);
	}
	
	open() {
		this.dialog.open();
	}
	
	close() {
		this.dialog.close();
	}
	
	destroy() {
		this.dialog.destroy();
	}
	
	setContent(content) {
		this.dialog.setContent(content);
	}
	
	setLocation(anchor) {
		this.dialog.setLocation(anchor);
	}
	
	setSize(size) {
		this.dialog.setSize(size);
	}
	
	lock() {
		this.dialog.lock();
	}
	
	unlock() {
		this.dialog.unlock();
	}
	
	freeze() {
		this.dialog.freeze();
	}
	
	unfreeze() {
		this.dialog.unfreeze();
	}
	
	hideClose() {
		this.dialog.hideClose();
	}
	
	showClose() {
		this.dialog.showClose();
	}
	
	hideResize() {
		this.dialog.hideResize();
	}
	
	showResize() {
		this.dialog.showResize();
	}

	render() {
		const { initOpen = true } = this.props;
		const options = {
			center: [0, 0],
			zoom: 1,
			minZoom: 1,
			maxZoom: 22,
		};
		return (
			<Map {...options}>
				<Dialog
					ref={(ref) => { this.dialog = ref; }}
					anchor={[50, 50]}
					id="testDialog"
					initOpen={initOpen}
				>
					{ '<div>This is the dialog content.</div>' }
				</Dialog>
			</Map>
		);
	}
}


src/Dialog.js
import { Children } from 'react';
import { render } from 'react-dom';
import { MapControl } from 'react-leaflet';
import L from 'leaflet';
import 'leaflet-dialog';
import 'leaflet-dialog/Leaflet.Dialog.css';

export default class Dialog extends MapControl {
	createLeafletElement(props) {
		this._id = props.id || this._generateId();
		this._containerId = `dialog-contents-${this._id}`;
		return L.control.dialog(props).setContent(`<div id="${this._containerId}"></div>`);
	}

	componentDidMount() {
		super.componentDidMount();
		const { map } = this.props.leaflet || this.context;
		const {
			onOpened,
			onClosed,
			onDestroyed,
			onLocked,
			onUnlocked,
			onFrozen,
			onUnfrozen,
			onUpdated,
			onResizeStart,
			onResizing,
			onResizeEnd,
			onMoveStart,
			onMoving,
			onMoveEnd,
			onCloseHidden,
			onCloseShown,
			onResizeHidden,
			onResizeShown,
			children
		} = this.props;
		this.leafletElement.addTo(map);
		this._renderContent(children);
		map.on('dialog:opened', (e) => {
				this._propagateEvent(onOpened, e);
			})
			.on('dialog:closed', (e) => {
				this._propagateEvent(onClosed, e);
			})
			.on('dialog:destroyed', (e) => {
				this._propagateEvent(onDestroyed, e);
			})
			.on('dialog:locked', (e) => {
				this._propagateEvent(onLocked, e);
			})
			.on('dialog:unlocked', (e) => {
				this._propagateEvent(onUnlocked, e);
			})
			.on('dialog:frozen', (e) => {
				this._propagateEvent(onFrozen, e);
			})
			.on('dialog:unfrozen', (e) => {
				this._propagateEvent(onUnfrozen, e);
			})
			.on('dialog:updated', (e) => {
				this._propagateEvent(onUpdated, e);
			})
			.on('dialog:resizestart', (e) => {
				this._propagateEvent(onResizeStart, e);
			})
			.on('dialog:resizing', (e) => {
				this._propagateEvent(onResizing, e);
			})
			.on('dialog:resizeend', (e) => {
				this._propagateEvent(onResizeEnd, e);
			})
			.on('dialog:movestart', (e) => {
				this._propagateEvent(onMoveStart, e);
			})
			.on('dialog:moving', (e) => {
				this._propagateEvent(onMoving, e);
			})
			.on('dialog:moveend', (e) => {
				this._propagateEvent(onMoveEnd, e);
			})
			.on('dialog:closehidden', (e) => {
				this._propagateEvent(onCloseHidden, e);
			})
			.on('dialog:closeshown', (e) => {
				this._propagateEvent(onCloseShown, e);
			})
			.on('dialog:resizehidden', (e) => {
				this._propagateEvent(onResizeHidden, e);
			})
			.on('dialog:resizeshown', (e) => {
				this._propagateEvent(onResizeShown, e);
			});
	}

	close() {
		this.leafletElement.close();
	}

	open() {
		this.leafletElement.open();
	}

	destroy() {
		this.leafletElement.destroy();
	}

	setContent(content) {
		this._renderContent(content);
	}

	setLocation(anchor) {
		this.leafletElement.setLocation(anchor);
	}

	setSize(size) {
		this.leafletElement.setSize(size);
	}

	lock() {
		this.leafletElement.lock();
	}

	unlock() {
		this.leafletElement.unlock();
	}

	freeze() {
		this.leafletElement.freeze();
	}

	unfreeze() {
		this.leafletElement.unfreeze();
	}

	hideClose() {
		this.leafletElement.hideClose();
	}

	showClose() {
		this.leafletElement.showClose();
	}

	hideResize() {
		this.leafletElement.hideResize();
	}

	showResize() {
		this.leafletElement.showResize();
	}

	_renderContent(content) {
		if (typeof content === 'string') {
			this.leafletElement.setContent(`<div id="${this._containerId}">${content}</div>`);
		} else {
			render(
				Children.only(content),
				document.getElementById(this._containerId)
			);
		}
	}

	_propagateEvent(eventHandler, event) {
		if (typeof eventHandler !== 'function') return;
		eventHandler(event);
	}

	_generateId() {
		return (Math.random() * 1e20).toString(36);
	}
}


webpack.config.js
/* eslint-disable */
module.exports = {
	entry: './dist/react-leaflet-dialog.min.js',
	output: {
		library: {
			root: 'ReactLeafletDialog',
			amd: 'react-leaflet-dialog',
			commonjs: 'react-leaflet-dialog'
		},
		libraryExport: 'default',
		libraryTarget: 'umd'
	},
	externals: {
		leaflet: {
			commonjs: 'leaflet',
			commonjs2: 'leaflet',
			root: 'L'
		},
		'react-leaflet': {
			commonjs: 'react-leaflet',
			commonjs2: 'react-leaflet',
			root: 'ReactLeaflet'
		},
		react: {
			commonjs: 'react',
			commonjs2: 'react',
			root: 'React'
		}
	},
	mode: 'production',
	module: {
		rules: [
			{
				test: /\.(js|jsx)$/,
                exclude: /(node_modules|bower_components)/,
                loader: 'babel-loader',
                options: {
                    presets: ['env', 'react'],
                    plugins: ['transform-class-properties', 'transform-object-rest-spread']
                }
			},
			{
				test: /\.css$/,
				exclude: /(node_modules)((?!leaflet-dialog).)*$/,
				use: [
					{ loader: 'style-loader' },
					{ loader: 'css-loader' },
					{ loader: 'postcss-loader' }
				]
			},
			{
				test: /\.(eot|woff|woff2|ttf|svg|png|jpe?g|gif)(\?\S*)?$/,
				exclude: /node_modules/,
				use: [
					{ loader: 'url?limit=100000&name=[name].[ext]' }
				]
			}
		]
	}
};



lib/Leaflet.Dialog.js
'use strict';

L.Control.Dialog = L.Control.extend({
  options: {
    size: [300, 300],
    minSize: [100, 100],
    maxSize: [350, 350],
    anchor: [250, 250],
    position: 'topleft',
    initOpen: true
  },

  initialize: function initialize(options) {
    this.options = JSON.parse(JSON.stringify(this.options));
    L.setOptions(this, options);

    this._attributions = {};
  },

  onAdd: function onAdd(map) {

    this._initLayout();
    this._map = map;

    this.update();

    if (!this.options.initOpen) {
      this.close();
    }

    return this._container;
  },

  open: function open() {
    if (!this._map) {
      return;
    }
    this._container.style.visibility = '';

    this._map.fire('dialog:opened', this);

    return this;
  },

  close: function close() {
    this._container.style.visibility = 'hidden';

    this._map.fire('dialog:closed', this);
    return this;
  },

  destroy: function destroy() {
    if (!this._map) {
      return this;
    }

    // this.removeFrom(this._map);
    // this.remove();

    if (this.onRemove) {
      this.onRemove(this._map);
    }

    this._map.fire('dialog:destroyed', this);
    this.remove();
    return this;
  },

  setLocation: function setLocation(location) {
    location = location || [250, 250];

    this.options.anchor[0] = 0;
    this.options.anchor[1] = 0;
    this._oldMousePos.x = 0;
    this._oldMousePos.y = 0;

    this._move(location[1], location[0]);

    return this;
  },

  setSize: function setSize(size) {
    size = size || [300, 300];

    this.options.size[0] = 0;
    this.options.size[1] = 0;
    this._oldMousePos.x = 0;
    this._oldMousePos.y = 0;

    this._resize(size[0], size[1]);

    return this;
  },

  lock: function lock() {
    this._resizerNode.style.visibility = 'hidden';
    this._grabberNode.style.visibility = 'hidden';
    this._closeNode.style.visibility = 'hidden';

    this._map.fire('dialog:locked', this);
    return this;
  },

  unlock: function unlock() {
    this._resizerNode.style.visibility = '';
    this._grabberNode.style.visibility = '';
    this._closeNode.style.visibility = '';

    this._map.fire('dialog:unlocked', this);
    return this;
  },

  freeze: function freeze() {
    this._resizerNode.style.visibility = 'hidden';
    this._grabberNode.style.visibility = 'hidden';

    this._map.fire('dialog:frozen', this);
    return this;
  },

  unfreeze: function unfreeze() {
    this._resizerNode.style.visibility = '';
    this._grabberNode.style.visibility = '';

    this._map.fire('dialog:unfrozen', this);
    return this;
  },

  hideClose: function hideClose() {
    this._closeNode.style.visibility = 'hidden';

    this._map.fire('dialog:closehidden', this);
    return this;
  },

  showClose: function showClose() {
    this._closeNode.style.visibility = '';

    this._map.fire('dialog:closeshown', this);
    return this;
  },

  hideResize: function hideResize() {
    this._resizerNode.style.visibility = 'hidden';

    this._map.fire('dialog:resizehidden', this);
    return this;
  },

  showResize: function showResize() {
    this._resizerNode.style.visibility = '';

    this._map.fire('dialog:resizeshown', this);
    return this;
  },

  setContent: function setContent(content) {
    this._content = content;
    this.update();
    return this;
  },

  getContent: function getContent() {
    return this._content;
  },

  getElement: function getElement() {
    return this._container;
  },

  update: function update() {
    if (!this._map) {
      return;
    }

    this._container.style.visibility = 'hidden';

    this._updateContent();
    this._updateLayout();

    this._container.style.visibility = '';
    this._map.fire('dialog:updated', this);
  },

  _initLayout: function _initLayout() {
    var className = 'leaflet-control-dialog',
        container = this._container = L.DomUtil.create('div', className);

    container.style.width = this.options.size[0] + 'px';
    container.style.height = this.options.size[1] + 'px';

    container.style.top = this.options.anchor[0] + 'px';
    container.style.left = this.options.anchor[1] + 'px';

    var stop = L.DomEvent.stopPropagation;
    L.DomEvent
    // .on(container, 'click', stop)
    .on(container, 'mousedown', stop).on(container, 'touchstart', stop).on(container, 'dblclick', stop).on(container, 'mousewheel', stop).on(container, 'contextmenu', stop).on(container, 'MozMousePixelScroll', stop);

    var innerContainer = this._innerContainer = L.DomUtil.create('div', className + '-inner');

    var grabberNode = this._grabberNode = L.DomUtil.create('div', className + '-grabber');
    var grabberIcon = L.DomUtil.create('i', 'fa fa-arrows');
    grabberNode.appendChild(grabberIcon);

    L.DomEvent.on(grabberNode, 'mousedown', this._handleMoveStart, this);

    var closeNode = this._closeNode = L.DomUtil.create('div', className + '-close');
    var closeIcon = L.DomUtil.create('i', 'fa fa-times');
    closeNode.appendChild(closeIcon);
    L.DomEvent.on(closeNode, 'click', this._handleClose, this);

    var resizerNode = this._resizerNode = L.DomUtil.create('div', className + '-resizer');
    var resizeIcon = L.DomUtil.create('i', 'fa fa-arrows-h fa-rotate-45');
    resizerNode.appendChild(resizeIcon);

    L.DomEvent.on(resizerNode, 'mousedown', this._handleResizeStart, this);

    var contentNode = this._contentNode = L.DomUtil.create('div', className + "-contents");

    container.appendChild(innerContainer);

    innerContainer.appendChild(contentNode);
    innerContainer.appendChild(grabberNode);
    innerContainer.appendChild(closeNode);
    innerContainer.appendChild(resizerNode);

    this._oldMousePos = { x: 0, y: 0 };
  },

  _handleClose: function _handleClose() {
    this.close();
  },

  _handleResizeStart: function _handleResizeStart(e) {
    this._oldMousePos.x = e.clientX;
    this._oldMousePos.y = e.clientY;

    L.DomEvent.on(this._map, 'mousemove', this._handleMouseMove, this);
    L.DomEvent.on(this._map, 'mouseup', this._handleMouseUp, this);

    this._map.fire('dialog:resizestart', this);
    this._resizing = true;
  },

  _handleMoveStart: function _handleMoveStart(e) {
    this._oldMousePos.x = e.clientX;
    this._oldMousePos.y = e.clientY;

    L.DomEvent.on(this._map, 'mousemove', this._handleMouseMove, this);
    L.DomEvent.on(this._map, 'mouseup', this._handleMouseUp, this);

    this._map.fire('dialog:movestart', this);
    this._moving = true;
  },

  _handleMouseMove: function _handleMouseMove(e) {
    var diffX = e.originalEvent.clientX - this._oldMousePos.x,
        diffY = e.originalEvent.clientY - this._oldMousePos.y;

    // this helps prevent accidental highlighting on drag:
    if (e.originalEvent.stopPropagation) {
      e.originalEvent.stopPropagation();
    }
    if (e.originalEvent.preventDefault) {
      e.originalEvent.preventDefault();
    }

    if (this._resizing) {
      this._resize(diffX, diffY);
    }

    if (this._moving) {
      this._move(diffX, diffY);
    }
  },

  _handleMouseUp: function _handleMouseUp() {
    L.DomEvent.off(this._map, 'mousemove', this._handleMouseMove, this);
    L.DomEvent.off(this._map, 'mouseup', this._handleMouseUp, this);

    if (this._resizing) {
      this._resizing = false;
      this._map.fire('dialog:resizeend', this);
    }

    if (this._moving) {
      this._moving = false;
      this._map.fire('dialog:moveend', this);
    }
  },

  _move: function _move(diffX, diffY) {
    var newY = this.options.anchor[0] + diffY;
    var newX = this.options.anchor[1] + diffX;

    this.options.anchor[0] = newY;
    this.options.anchor[1] = newX;

    this._container.style.top = this.options.anchor[0] + 'px';
    this._container.style.left = this.options.anchor[1] + 'px';

    this._map.fire('dialog:moving', this);

    this._oldMousePos.y += diffY;
    this._oldMousePos.x += diffX;
  },

  _resize: function _resize(diffX, diffY) {
    var newX = this.options.size[0] + diffX;
    var newY = this.options.size[1] + diffY;

    if (newX <= this.options.maxSize[0] && newX >= this.options.minSize[0]) {
      this.options.size[0] = newX;
      this._container.style.width = this.options.size[0] + 'px';
      this._oldMousePos.x += diffX;
    }

    if (newY <= this.options.maxSize[1] && newY >= this.options.minSize[1]) {
      this.options.size[1] = newY;
      this._container.style.height = this.options.size[1] + 'px';
      this._oldMousePos.y += diffY;
    }

    this._map.fire('dialog:resizing', this);
  },

  _updateContent: function _updateContent() {

    if (!this._content) {
      return;
    }

    var node = this._contentNode;
    var content = typeof this._content === 'function' ? this._content(this) : this._content;

    if (typeof content === 'string') {
      node.innerHTML = content;
    } else {
      while (node.hasChildNodes()) {
        node.removeChild(node.firstChild);
      }
      node.appendChild(content);
    }
  },

  _updateLayout: function _updateLayout() {

    this._container.style.width = this.options.size[0] + 'px';
    this._container.style.height = this.options.size[1] + 'px';

    this._container.style.top = this.options.anchor[0] + 'px';
    this._container.style.left = this.options.anchor[1] + 'px';
  }

});

L.control.dialog = function (options) {
  return new L.Control.Dialog(options);
};